llvm/lib/Transforms/IPO/FunctionSpecialization.cpp

```diff
@@ -262,29 +269,113 @@ Cost InstCostVisitor::estimateBranchInst(BranchInst &I) {
   return estimateBasicBlocks(WorkList);
 }
 
+bool InstCostVisitor::discoverTransitivelyIncomingValues(
+    Constant *Const, PHINode *Root, DenseSet<PHINode *> &TransitivePHIs,
+    SmallVectorImpl<PHINode *> &UnknownIncomingValues) {
+
+  SmallVector<PHINode *, 64> WorkList;
+  WorkList.push_back(Root);
+  unsigned Iter = 0;
+
+  while (!WorkList.empty()) {
+    PHINode *PN = WorkList.pop_back_val();
+
+    if (++Iter > MaxDiscoveryIterations ||
+        PN->getNumIncomingValues() > MaxIncomingPhiValues) {
+      // For now just collect the Phi and later we will check whether it is
+      // in the Transitive set.
+      UnknownIncomingValues.push_back(PN);
+      continue;
+      // FIXME: return false here and remove the UnknownIncomingValues entirely.
```

Comment:
Let's implement FIXME actually. It might not be hard.

Reply:
This will require the regresson test to be adjusted for the NOFUNCSPEC check prefix. Last time I checked it would require another 55 discovery steps or so. But that's fine because we will be removing the following loop from visitPHI, which iterates over these 55 UnknownIncomingValues.
```
for (PHINode *Phi : UnknownIncomingValues)
    if (!TransitivePHIs.contains(Phi))
      return nullptr;
```